composer/datasets/ade20k.py

```diff
@@ -388,6 +391,136 @@ def initialize_object(self, batch_size, dataloader_hparams) -> DataSpec:
+        x = obj['image']
+        y = obj['annotation']
+        if self.both_transform:
+            x, y = self.both_transform((x, y))
+        if self.image_transform:
+            x = self.image_transform(x)
+        if self.annotation_transform:
+            y = self.annotation_transform(y)
+        return x, y
+
+
+@dataclass
+class StreamingADE20kHparams(DatasetHparams):
+    """Streaming ADE20k hyperparameters.
+
+    Args:
+        remote (str): Remote directory (S3 or local filesystem) where dataset is stored.
+        local (str): Local filesystem directory where dataset is cached during operation.
+        split (str): the dataset split to use either 'train', 'val', or 'test'. Default: ``'train```.
+        base_size (int): initial size of the image and target before other augmentations. Default: ``512``.
+        min_resize_scale (float): the minimum value the samples can be rescaled. Default: ``0.5``.
+        max_resize_scale (float): the maximum value the samples can be rescaled. Default: ``2.0``.
+        final_size (int): the final size of the image and target. Default: ``512``.
+        ignore_background (bool): if true, ignore the background class when calculating the training loss.
+            Default: ``true``.
+    """
+
+    remote: str = hp.optional('Remote directory (S3 or local filesystem) where dataset is stored',
+                              default='s3://mosaicml-internal-dataset-ade20k/mds/')
+    local: str = hp.optional('Local filesystem directory where dataset is cached during operation',
+                             default='/tmp/mds-cache/mds-ade20k/')
+    split: str = hp.optional("Which split of the dataset to use. Either ['train', 'val', 'test']", default='train')
+
+    base_size: int = hp.optional("Initial size of the image and target before other augmentations", default=512)
+    min_resize_scale: float = hp.optional("Minimum value that the image and target can be scaled", default=0.5)
+    max_resize_scale: float = hp.optional("Maximum value that the image and target can be scaled", default=2.0)
+    final_size: int = hp.optional("Final size of the image and target", default=512)
+    ignore_background: bool = hp.optional("If true, ignore the background class in training loss", default=True)
+
+    def validate(self):
+        if self.split not in ['train', 'val', 'test']:
+            raise ValueError(f"split value {self.split} must be one of ['train', 'val', 'test'].")
+
+        if self.base_size <= 0:
+            raise ValueError("base_size cannot be zero or negative.")
+
+        if self.min_resize_scale <= 0:
+            raise ValueError("min_resize_scale cannot be zero or negative")
+
+        if self.max_resize_scale < self.min_resize_scale:
+            raise ValueError("max_resize_scale cannot be less than min_resize_scale")
+
+    def initialize_object(self, batch_size: int, dataloader_hparams: DataLoaderHparams) -> DataSpec:
+        self.validate()
```

Comment:
Yes can do... I have a question though, do we want this preprocessing functionality like `split`, `base_size`, `min_resize_scale`, ... to all be part of our `StreamingADE20k` dataset? I'm in favor of this, just want to emphasize that now it won't just be an ADE20k dataset that returns PIL images, it's a dataset with some built in user-chosen preprocessing steps. So less reusable code, but more clearly an _example_ of how a user constructs their custom dataset.

Reply:
Also, dataloader construction will still live in `initialize_object` by necessity, because `dataloader_hparams` and `batch_size` are passed in via the Trainer